File: examples/kmm/shared/src/commonMain/kotlin/com/moriatsushi/koject/example/kmm/viewmodel/TodoListViewModel.kt
package com.moriatsushi.koject.example.kmm.viewmodel

import com.moriatsushi.koject.example.kmm.model.TodoList
import com.moriatsushi.koject.example.kmm.model.TodoTask
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.StateFlow

interface TodoListViewModel {
    val coroutineScope: CoroutineScope
    val list: StateFlow<TodoList>

    fun addTask(title: String)
    fun changeComplete(task: TodoTask, isCompleted: Boolean)
    fun onCleared()
}

File: examples/kmm/ios/ios/viewmodel/TodoListViewModel.swift
import Combine
import shared

import Foundation

class TodoListViewModel: ObservableObject {
    @Published
    var list: DataTodoList = DataTodoList.Companion().empty
    
    private let viewModel: UiIOSTodoListViewModel =
        KojectHelper_iosKt.injectTodoListViewModel()
    
    init() {
        self.viewModel.observeTasks { [unowned self] list in
            self.list = list
        }
    }
    
    func addTask(title: String) {
        self.viewModel.addTask(title: title)
    }
    
    func changeComplete(task: DataTodoTask, isCompleted: Bool) {
        self.viewModel.changeComplete(task: task, isCompleted: isCompleted)
    }
    
    deinit {
        self.viewModel.onCleared()
    }
}
